lib/task.ml

```diff
@@ -1,33 +1,102 @@
+open EffectHandlers
+open EffectHandlers.Deep
+
 type 'a task = unit -> 'a
 
-type 'a promise = ('a, exn) result option Atomic.t
+type 'a promise_state =
+  Returned of 'a
+| Raised of exn
+| Pending of (unit, unit) continuation list
 
-exception TasksActive
+type 'a promise = 'a promise_state Atomic.t
 
-type task_msg =
-  Task : 'a task * 'a promise -> task_msg
-| Quit : task_msg
+type message =
+  Work of (unit -> unit)
+| Quit
 
 type pool_data = {
   domains : unit Domain.t array;
-  task_chan : task_msg Multi_channel.t;
+  task_chan : message Multi_channel.t;
   name: string option
 }
 
 type pool = pool_data option Atomic.t
 
-let do_task f p =
+let get_pool_data p =
+  match Atomic.get p with
+  | None -> raise (Invalid_argument "pool already torn down")
+  | Some p -> p
+
+let do_task pd f p =
   try
     let res = f () in
-    Atomic.set p (Some (Ok res))
+    match Atomic.exchange p (Returned res) with
```

Comment:
could the 2 branches here be factored out like this? (in the common case it might avoid allocating a closure for the List.iter param)
```ocaml
let cont k = Multi_channel.send pd.task_chan (Work (continue k))
let discont e k = Multi_channel.send pd.task_chan (Work (fun _ -> discontinue k e))

let do_task pd f p =
  let action, result = try  cont, Returned  (f ()) with e -> discont e, Raised e in
  match Atomic.exchange p result with
  | Pending l -> List.iter action l
  |  _ -> failwith "Task.do_task: impossible, can only set result of task once"
```

Reply:
Also as an optimization this could pick up the first task, queue up the rest with multi-channel and then immediately run the first task on success?

Reply:
Refactored code. 

Reg optimisation, `Multi_channel` is a work-stealing queue and optimises for local pops. We should come back to this optimisation if there is need for it. Moreover, I'd like `do_task` to be able to return to its caller so that `run` gets a chance to examine the promise for completion. All of this is certainly doable, and we should come back to it later.